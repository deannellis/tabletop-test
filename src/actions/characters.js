import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

