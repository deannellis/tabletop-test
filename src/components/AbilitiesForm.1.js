import React from 'react';

export default class AbilitiesForm extends React.Component {
    state = {
        strength: undefined,
        intelligence: undefined,
        wisdom: undefined,
        dexterity: undefined,
        constitution: undefined,
        charisma: undefined
    };
    
    renderButtons = () => {
        Object.keys(this.state).map((key) => {
            return 
            <div>
                <button type="button" onClick={this.onRollStrength} disabled={typeof this.state.strength != 'undefined'}>Strength</button>
                <p>Strength: {typeof this.state.strength == 'undefined' ? 'Roll to determine Strength' : this.state.strength }</p>
            
        });
    };
    
    onRollStrength = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({strength: rollValue}));
    };
    
    onRollIntelligence = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({intelligence: rollValue}));
    };
    
    onRollWisdom = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({wisdom: rollValue}));
    };
    
    onRollDexterity = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({dexterity: rollValue}));
    };
    
    onRollConstitution = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({constitution: rollValue}));
    };
    
    render() {
        return (
            <div>
                <form>
                    {this.renderButtons()}
                </form>
            </div>
        );
    }
}