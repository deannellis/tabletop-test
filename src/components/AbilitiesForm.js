import React from 'react';
import SingleRollDie from './SingleRollDie';
import uuid from 'uuid';

export default class AbilitiesForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            strength: [],
            intelligence: [],
            wisdom: [],
            dexterity: [],
            constitution: [],
            charisma: [],
            error: '',
            refreshKey: 0,
            lastDieRolled: undefined,
            lastResult: undefined
        };
    }
    
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    {this.renderAttribute('Strength', 'strength')}
                    {this.renderAttribute('Intelligence', 'intelligence')}
                    {this.renderAttribute('Wisdom', 'wisdom')}
                    {this.renderAttribute('Dexterity', 'dexterity')}
                    {this.renderAttribute('Constitution', 'constitution')}
                    {this.renderAttribute('Charisma', 'charisma')}
                    <button>Next Step</button>
                </form>
            </div>
        );
    }
    
    handleRollDie = (result, attributeKey) => {
        this.setState(() => ({
            lastResult: result,
            [attributeKey]: [...this.state[attributeKey], result]
        }));
    };
    
    onSubmit = (e) => {
        e.preventDefault();
        
        const sum = (accumulator, currentValue) => accumulator + currentValue;
        
        if(this.state.strength.length < 3 || this.state.intelligence.length < 3 || this.state.wisdom.length < 3 || this.state.dexterity.length < 3 || this.state.constitution.length < 3 || this.state.charisma.length < 3) {
            this.setState(() => ({
                error: 'Please roll all scores'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit({
                id: uuid(),
                abilities: {
                    strength: this.state.strength.reduce(sum),
                    intelligence: this.state.intelligence.reduce(sum),
                    wisdom: this.state.wisdom.reduce(sum),
                    dexterity: this.state.dexterity.reduce(sum),
                    constitution: this.state.constitution.reduce(sum),
                    charisma: this.state.charisma.reduce(sum)
                },
                name: 'In-Progress Character',
                charClass: 'unkown'
            });
        } 
    }
    
    renderAttribute = (label, attributeKey) => {
        const attributeValue = this.state[attributeKey];
        const isDefined = attributeValue.length !== 0;
        let rollsString = '';
        if(this.state[attributeKey].length < 3){
            this.state[attributeKey].forEach(
                (roll) => {rollsString += `${roll} + `}
            );
        } else {
            let total = 0;
            this.state[attributeKey].forEach(
                (roll) => {total += roll}
            );
            rollsString = `${this.state[attributeKey][0]} + ${this.state[attributeKey][1]} + ${this.state[attributeKey][2]} = ${total}`;
        }
        
        const message = label + ': ' + (isDefined ? rollsString : 'roll to determine');
        return(
            <div>
                <p>{message}</p>
                <div className="dice-container">
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                </div>
            </div>
        );
    }
    
    updateRefreshKey = (newKey) => {
        this.setState(() => ({
            refreshKey: newKey
        }));
    };
}