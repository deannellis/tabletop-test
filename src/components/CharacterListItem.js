import React from 'react';
import { Link } from 'react-router-dom';

const CharacterListItem = ({ name, charClass, race, id }) => (
    <div>
        <h2>{name}</h2>
        <p>{race} - {charClass}</p>
        <Link to={`/view-char/${id}`}>View Character</Link>
    </div>    
);

export default CharacterListItem;
