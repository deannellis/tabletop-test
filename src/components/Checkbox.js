import React, { Component, PropTypes } from 'react';

class Checkbox extends React.Component {
    state = {
        isChecked: false,
    }
    
    toggleCheckboxChange = () => {
        const { handleCheckboxChange } = this.props;
        const { isChecked } = this.state;
    
        this.setState(({ isChecked }) => (
            {
                isChecked: !isChecked,
            }
        ));
    
        handleCheckboxChange(!isChecked);
    }
    
      render() {
        const { label } = this.props;
        const { isChecked } = this.state;
        let isDisabled;
        if(isChecked) {
            isDisabled = false;
        } else {
            isDisabled = this.props.isDisabled
        }
    
        return (
          <div className="checkbox">
            <label>
                <input
                    type="checkbox"
                    value={isChecked}
                    checked={isChecked}
                    onChange={this.toggleCheckboxChange}
                    disabled={isDisabled}
                />
    
              {label}
            </label>
          </div>
        );
      }
}

export default Checkbox;