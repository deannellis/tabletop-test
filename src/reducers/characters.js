//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};