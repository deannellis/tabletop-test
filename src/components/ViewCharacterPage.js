import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import IdListItem from './IdListItem';
import NotFoundPage from './NotFoundPage';
import { removeCharacter } from '../actions/characters';

export class ViewCharacterPage extends React.Component {
    
    constructor(props){
        super(props);
        this.character = this.props.character;
    }
    
    handleRemoveCharacter = () => {
        this.props.handleRemoveCharacter(this.character.id);
        this.props.history.push('/');
    };
    
    render() {
        if(this.props.character !== undefined) {
            return (
                <div>
                    <h1>{this.character.name}</h1>
                    {this.character.description && <p>{this.character.description}</p>}
                    <p>{this.character.race} - {this.character.charClass}</p>
                    <h2>Ability Scores</h2>
                    <p>{`Strength: ${this.character.abilities.strength}`}</p>
                    <p>{`Intelligence: ${this.character.abilities.intelligence}`}</p>
                    <p>{`Wisdom: ${this.character.abilities.wisdom}`}</p>
                    <p>{`Dexterity: ${this.character.abilities.dexterity}`}</p>
                    <p>{`Charisma: ${this.character.abilities.charisma}`}</p>
                    <p>{`Constitution: ${this.character.abilities.constitution}`}</p>
                    <p>{`HP: ${this.character.hp}`}</p>
                    <p>{`XP: ${this.character.xp}`}</p>
                    <p>{`Gold: ${this.character.gold}`}</p>
                    {this.character.hasOwnProperty("spells") && 
                        <div>
                            <h3>Spells:</h3>
                            <div>
                                {this.character.spells.map((id) => {
                                    return <IdListItem 
                                                key={id} 
                                                id={id}
                                                obj={'spells'}
                                            />
                                })}
                            </div>
                        </div>
                    }
                    <h3>Equipmunk:</h3>
                    <div>
                        {this.character.equipment.map((id) => {
                            return <IdListItem 
                                        key={id} 
                                        id={id}
                                        obj={'weapons'}
                                    />
                        })}
                    </div>
                    <Link to="/">Back</Link>
                    <div>
                        <button onClick={this.handleRemoveCharacter}>Delete Character</button>
                    </div>
                </div>    
            );
        } else {
            return (
                <NotFoundPage />
            );
        }
    }
}

const mapStateToProps = (state, props) => ({
    character: state.characters.find((character) => character.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    handleRemoveCharacter: (charId) => dispatch(removeCharacter({ id: charId }))
});

export default connect(mapStateToProps, mapDispatchToProps)(ViewCharacterPage);