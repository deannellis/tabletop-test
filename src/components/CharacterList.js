import React from 'react';
import { connect } from 'react-redux';
import CharacterListItem from './CharacterListItem';

const CharacterList = (props) => (
    <div>
        <h1>Character List</h1>
        {props.characters.length > 0 ?
            <div>
                {props.characters.map((character) => {
                    return <CharacterListItem key={character.id} {...character} />
                })}
            </div>
        :
            <p>You have no characters, Click below to create a character</p>
        }
    </div>    
);

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(CharacterList);
