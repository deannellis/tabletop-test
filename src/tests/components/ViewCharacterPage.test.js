/* global jest expect */
import React from 'react';
import { shallow } from 'enzyme';
import { ViewCharacterPage } from '../../components/ViewCharacterPage';
import characters from '../fixtures/characters';

test('should render view character page without props', () => {
    const wrapper = shallow(<ViewCharacterPage />);
    expect(wrapper).toMatchSnapshot();
});

test('should render view character page with props', () => {
    const wrapper = shallow(<ViewCharacterPage character={characters[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should call handleRemoveCharacter', () => {
    const handleRemoveCharacter = jest.fn();
    const history = { push: jest.fn() }
    const wrapper = shallow(<ViewCharacterPage 
                                handleRemoveCharacter={handleRemoveCharacter}
                                character={characters[0]}
                                history={history}
                            />);
    wrapper.find('button').simulate('click');
    expect(handleRemoveCharacter).toHaveBeenLastCalledWith(characters[0].id);
    expect(history.push).toHaveBeenLastCalledWith('/');
});