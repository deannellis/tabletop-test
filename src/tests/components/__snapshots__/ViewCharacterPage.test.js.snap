// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render view character page with props 1`] = `
<div>
  <h1>
    Rick Slick
  </h1>
  <p>
    Loves his Mom
  </p>
  <p>
    human
     - 
    magic-user
  </p>
  <h2>
    Ability Scores
  </h2>
  <p>
    Strength: 11
  </p>
  <p>
    Intelligence: 14
  </p>
  <p>
    Wisdom: 11
  </p>
  <p>
    Dexterity: 9
  </p>
  <p>
    Charisma: 9
  </p>
  <p>
    Constitution: 9
  </p>
  <p>
    HP: 1
  </p>
  <p>
    XP: 0
  </p>
  <p>
    Gold: 109
  </p>
  <div>
    <h3>
      Spells:
    </h3>
    <div>
      <Component
        id="1"
        key="1"
        obj="spells"
      />
    </div>
  </div>
  <h3>
    Equipmunk:
  </h3>
  <div>
    <Component
      id={0}
      key="0"
      obj="weapons"
    />
    <Component
      id={1}
      key="1"
      obj="weapons"
    />
  </div>
  <Link
    replace={false}
    to="/"
  >
    Back
  </Link>
  <div>
    <button
      onClick={[Function]}
    >
      Delete Character
    </button>
  </div>
</div>
`;

exports[`should render view character page without props 1`] = `<NotFoundPage />`;
